src/Algebra/Graph/Acyclic/Labelled/Algorithm.hs

```diff
@@ -0,0 +1,35 @@
+module Algebra.Graph.Acyclic.Labelled.Algorithm where
+
+import Algebra.Graph.Acyclic.Labelled.AdjacencyMap
+import qualified Algebra.Graph.AdjacencyMap.Algorithm as AM
+import Algebra.Graph.Label
+import qualified Algebra.Graph.Labelled.AdjacencyMap as LAM
+import Data.Map.Strict (Map, (!))
+import qualified Data.Map.Strict as Map
+import Data.Maybe (fromMaybe)
+
+-- TODO: Replace this function with 'skeleton' from Acyclic.Labelled to remove the use of fromMaybe
+-- TODO: Make 'topSort' more efficient
+-- TODO: Add examples and tests
+topSort :: (Ord a) => AdjacencyMap e a -> [a]
+topSort = fromMaybe [] . AM.topSort . LAM.skeleton . fromAcyclic
+
+-- TODO: Add time complexity
+-- TODO: Add examples using 'Optimum' data type
+-- | Compute the /shortest path/ to each vertex in the graph
+-- from a given source vertex.
+--
+-- The following examples assume that the edges are distances,
+-- ie. the edge 'Semiring' is 'Distance'.
+-- @
+-- dijkstra ('LAM.toAcyclicOrd' $ 'LAM.edges' [(2, 'b', 'c'), (1, 'a', 'b'), (3, 'a', 'c')]) 'z' == Map.'Map.fromList' [('z', 0), ('a', Infinite), ('b', Infinite), ('c', Infinite)]
+-- dijkstra ('LAM.toAcyclicOrd' $ 'LAM.edges' [(2, 'b', 'c'), (1, 'a', 'b'), (3, 'a', 'c')]) 'a' == Map.'Map.fromList' [('a', 0), ('b', 1), ('c', 3)]
+-- @
+optimumPath :: (Semiring e, Ord a) => AdjacencyMap e a -> a -> Map a e
+optimumPath am s = foldl (relaxVertex em) (initialize em) vL
+  where
+    vL = dropWhile (/=s) $ topSort am
+    em = (LAM.adjacencyMap . fromAcyclic) am
+    initialize = Map.insert s one . Map.map (const zero)
```

Comment:
If the `s` doesn't belong to the graph you still mark it with `one` in the result. I'm not sure this is the right behaviour.

Reply:
I agree I should probably change the result to `Maybe (Map a e)` as suggested.

Reply:
I'm not suggesting to go with `Maybe (Map a e)` as the final result, but it might appear somewhere internally. I am just saying we should think carefully about what we return in the case `s` is not in the graph. Is `s -> one` what we want? Maybe `s -> zero`? Or perhaps even `Nothing` but that may be inconvenient.

Reply:
Perhaps, the right thing to do is to return a map that doesn't have an entry for `s` in this case.

Do we guarantee that the resulting map contains an entry for each of the graph vertices?

Reply:
> Do we guarantee that the resulting map contains an entry for each of the graph vertices?

Yes.

> Perhaps, the right thing to do is to return a map that doesn't have an entry for s in this case.

I think to return `Nothing` is more efficient (as I am using `fmap`) and meaningful in this case.

Reply:
If using `Nothing` is more efficient you can use in the implementation, but you don't necessarily need to use it in the interface too.

As for being meaningful, I think using a `const zero` map is more meaningful. Why? Because you are saying that all graph vertices are unreachable from the given vertex, which is correct. With `Maybe` you can use `Nothing` to mean the same thing, but then `Just zeroMap` will be an impossible and unused output (because if the source vertex belongs to the graph then it will have `one` associated with it). As a result, I think using `Maybe` doesn't bring any type-safety, it's merely a choice between using `Nothing` and `zeroMap` to represent the same thing. Using `Maybe` will probably be just less convenient in practice.